**<참고>**
- 기상 데이터 및 황사(미세먼지) 데이터는 **"서울" 지역**에 해당하는 자료들로만 구성하였다.

# 기상 데이터 작업
- 사용할 변수들
    - 기온
    - 강수량
    - 적설량

In [1]:
import numpy as np
import pandas as pd
import datetime as dt
from tqdm.notebook import tqdm

In [2]:
weather = pd.read_csv('C:/Users/Playdata/2020_Bigcontest/dataset/weather/2019_weather.csv', encoding = 'cp949')
weather

지점 지점명                일시  기온(°C)  기온 QC플래그  강수량(mm)  강수량 QC플래그  \
0     108  서울  2019-01-01 00:00    -5.5       0.0      NaN        NaN   
1     108  서울  2019-01-01 01:00    -5.9       0.0      NaN        NaN   
2     108  서울  2019-01-01 02:00    -6.5       0.0      NaN        NaN   
3     108  서울  2019-01-01 03:00    -6.9       0.0      NaN        NaN   
4     108  서울  2019-01-01 04:00    -7.2       0.0      NaN        NaN   
...   ...  ..               ...     ...       ...      ...        ...   
8756  108  서울  2019-12-31 20:00    -7.0       NaN      NaN        9.0   
8757  108  서울  2019-12-31 21:00    -7.4       NaN      NaN        9.0   
8758  108  서울  2019-12-31 22:00    -8.2       NaN      NaN        9.0   
8759  108  서울  2019-12-31 23:00    -7.4       NaN      NaN        9.0   
8760  108  서울  2020-01-01 00:00    -6.5       NaN      0.0        NaN   

      풍속(m/s)  풍속 QC플래그  풍향(16방위)  ...  최저운고(100m )  시정(10m)  지면상태(지면상태코드)  \
0         1.0       0.0     290.0  ...          NaN     1573           NaN   
1         1.8       NaN     270.0  ...          NaN     1569           NaN   
2         1.2       NaN     290.0  ...          NaN     1462           NaN   
3         2.2       NaN     270.0  ...          NaN     1281           NaN   
4         1.2       NaN     270.0  ...          NaN     1195           NaN   
...       ...       ...       ...  ...          ...      ...           ...   
8756      2.9       NaN     290.0  ...          NaN     1912           NaN   
8757      0.8       NaN      20.0  ...          NaN     1993           NaN   
8758      0.0       NaN       0.0  ...          NaN     1990           NaN   
8759      0.0       NaN       0.0  ...         14.0     2000           NaN   
8760      0.0       NaN       0.0  ...          9.0     2000           NaN   

      현상번호(국내식)  지면온도(°C)  지면온도 QC플래그  5cm 지중온도(°C)  10cm 지중온도(°C)  \
0           NaN      -6.0         0.0          -3.4           -1.1   
1           NaN      -6.5         0.0          -3.9           -1.4   
2           NaN      -7.1         0.0          -4.2           -1.6   
3           NaN      -7.3         0.0          -4.6           -1.8   
4           NaN      -7.5         0.0          -4.7           -2.0   
...         ...       ...         ...           ...            ...   
8756        NaN      -4.1         NaN          -0.7            0.9   
8757        NaN      -4.4         NaN          -0.8            0.9   
8758        NaN      -4.6         NaN          -1.0            0.8   
8759        NaN      -3.5         NaN          -1.1            0.8   
8760        5.0      -2.8         NaN          -0.8            0.7   

      20cm 지중온도(°C)  30cm 지중온도(°C)  
0               0.0            1.0  
1               0.0            1.0  
2               0.0            1.0  
3               0.0            1.0  
4              -0.1            1.0  
...             ...            ...  
8756            2.3            3.3  
8757            2.3            3.3  
8758            2.2            3.2  
8759            2.2            3.2  
8760            2.1            3.2  

[8761 rows x 37 columns]

In [3]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8761 entries, 0 to 8760
Data columns (total 37 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   지점             8761 non-null   int64  
 1   지점명            8761 non-null   object 
 2   일시             8761 non-null   object 
 3   기온(°C)         8761 non-null   float64
 4   기온 QC플래그       351 non-null    float64
 5   강수량(mm)        824 non-null    float64
 6   강수량 QC플래그      1491 non-null   float64
 7   풍속(m/s)        8754 non-null   float64
 8   풍속 QC플래그       6 non-null      float64
 9   풍향(16방위)       8754 non-null   float64
 10  풍향 QC플래그       6 non-null      float64
 11  습도(%)          8761 non-null   float64
 12  습도 QC플래그       322 non-null    float64
 13  증기압(hPa)       8761 non-null   float64
 14  이슬점온도(°C)      8761 non-null   float64
 15  현지기압(hPa)      8761 non-null   float64
 16  현지기압 QC플래그     322 non-null    float64
 17  해면기압(hPa)      8761 non-null   float64
 18  해면기압 QC플

In [4]:
weatehr_var_select = weather[['일시', '기온(°C)', '강수량(mm)', '습도(%)', '적설(cm)']]
weatehr_var_select

일시  기온(°C)  강수량(mm)  습도(%)  적설(cm)
0     2019-01-01 00:00    -5.5      NaN   54.0     NaN
1     2019-01-01 01:00    -5.9      NaN   56.0     NaN
2     2019-01-01 02:00    -6.5      NaN   60.0     NaN
3     2019-01-01 03:00    -6.9      NaN   62.0     NaN
4     2019-01-01 04:00    -7.2      NaN   57.0     NaN
...                ...     ...      ...    ...     ...
8756  2019-12-31 20:00    -7.0      NaN   41.0     NaN
8757  2019-12-31 21:00    -7.4      NaN   43.0     NaN
8758  2019-12-31 22:00    -8.2      NaN   42.0     NaN
8759  2019-12-31 23:00    -7.4      NaN   45.0     NaN
8760  2020-01-01 00:00    -6.5      0.0   38.0     NaN

[8761 rows x 5 columns]

- 한글로 된 변수명들을 영어로 변경해주겠다.
    - 일시: date
    - 기온(°C): temp
    - 강수량(mm): rain
    - 습도(%): humidity
    - 적설(cm): snow

In [5]:
weatehr_var_select.rename(columns = {'일시':'date', '기온(°C)':'temp', '강수량(mm)':'rain', '습도(%)':'humidity', '적설(cm)':'snow'},
                          inplace = True)
weatehr_var_select

C:\Users\Playdata\anaconda3\lib\site-packages\pandas\core\frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


date  temp  rain  humidity  snow
0     2019-01-01 00:00  -5.5   NaN      54.0   NaN
1     2019-01-01 01:00  -5.9   NaN      56.0   NaN
2     2019-01-01 02:00  -6.5   NaN      60.0   NaN
3     2019-01-01 03:00  -6.9   NaN      62.0   NaN
4     2019-01-01 04:00  -7.2   NaN      57.0   NaN
...                ...   ...   ...       ...   ...
8756  2019-12-31 20:00  -7.0   NaN      41.0   NaN
8757  2019-12-31 21:00  -7.4   NaN      43.0   NaN
8758  2019-12-31 22:00  -8.2   NaN      42.0   NaN
8759  2019-12-31 23:00  -7.4   NaN      45.0   NaN
8760  2020-01-01 00:00  -6.5   0.0      38.0   NaN

[8761 rows x 5 columns]

# 황사(미세먼지) 데이터 작업

In [6]:
dust = pd.read_csv('C:/Users/Playdata/2020_Bigcontest/dataset/weather/2019_dust.csv', encoding = 'cp949')
dust

지점 지점명                일시  1시간평균 미세먼지농도(㎍/㎥)
0     108  서울  2019-01-01 00:00                 59
1     108  서울  2019-01-01 01:00                 58
2     108  서울  2019-01-01 02:00                 44
3     108  서울  2019-01-01 03:00                 48
4     108  서울  2019-01-01 04:00                 51
...   ...  ..               ...                ...
8195  108  서울  2019-12-31 20:00                 27
8196  108  서울  2019-12-31 21:00                 27
8197  108  서울  2019-12-31 22:00                 29
8198  108  서울  2019-12-31 23:00                 24
8199  108  서울  2020-01-01 00:00                 27

[8200 rows x 4 columns]

In [7]:
dust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8200 entries, 0 to 8199
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   지점                 8200 non-null   int64 
 1   지점명                8200 non-null   object
 2   일시                 8200 non-null   object
 3   1시간평균 미세먼지농도(㎍/㎥)  8200 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 256.4+ KB


In [8]:
dust_var_select = dust[['일시', '1시간평균 미세먼지농도(㎍/㎥)']]
dust_var_select

일시  1시간평균 미세먼지농도(㎍/㎥)
0     2019-01-01 00:00                 59
1     2019-01-01 01:00                 58
2     2019-01-01 02:00                 44
3     2019-01-01 03:00                 48
4     2019-01-01 04:00                 51
...                ...                ...
8195  2019-12-31 20:00                 27
8196  2019-12-31 21:00                 27
8197  2019-12-31 22:00                 29
8198  2019-12-31 23:00                 24
8199  2020-01-01 00:00                 27

[8200 rows x 2 columns]

- 한글로 된 변수명들을 영어로 변경해주겠다.
    - 일시: date
    - 1시간평균 미세먼지농도(㎍/㎥): dust

In [9]:
dust_var_select.rename(columns = {'일시':'date', '1시간평균 미세먼지농도(㎍/㎥)':'dust'}, inplace = True)
dust_var_select

C:\Users\Playdata\anaconda3\lib\site-packages\pandas\core\frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


date  dust
0     2019-01-01 00:00    59
1     2019-01-01 01:00    58
2     2019-01-01 02:00    44
3     2019-01-01 03:00    48
4     2019-01-01 04:00    51
...                ...   ...
8195  2019-12-31 20:00    27
8196  2019-12-31 21:00    27
8197  2019-12-31 22:00    29
8198  2019-12-31 23:00    24
8199  2020-01-01 00:00    27

[8200 rows x 2 columns]

# 작업해 준 기상 데이터와 황사(미세먼지) 데이터를 결합

- 결합하기 전에 먼저 두 데이터 셋에 존재하는 date 변수(조인키)의 타입을 확인해봐야 한다.
    - 불일치할 경우, join이 제대로 수행되지 않기 때문에 확인 필수!!

In [10]:
weatehr_var_select.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8761 entries, 0 to 8760
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      8761 non-null   object 
 1   temp      8761 non-null   float64
 2   rain      824 non-null    float64
 3   humidity  8761 non-null   float64
 4   snow      82 non-null     float64
dtypes: float64(4), object(1)
memory usage: 342.4+ KB


In [11]:
dust_var_select.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8200 entries, 0 to 8199
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   date    8200 non-null   object
 1   dust    8200 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 128.2+ KB


- weatehr_var_select 데이터를 기준으로 두 데이터 셋을 **left join** 해 줄 것이다.
    - Left join을 해주는 이유는 weatehr_var_select 데이터 셋의 길이는 **8761개**인데, dust_var_select 데이터 셋의 길이는 **8200개**이다.
    - 따라서 더 많은 데이터가 존재하는 weatehr_var_select 데이터 셋을 기준으로 left join 해주는 것이다.
    - 조인키는 date 변수를 사용하겠다.

In [12]:
weather_final = pd.merge(weatehr_var_select, dust_var_select, how = 'left', on = 'date')
weather_final

date  temp  rain  humidity  snow  dust
0     2019-01-01 00:00  -5.5   NaN      54.0   NaN  59.0
1     2019-01-01 01:00  -5.9   NaN      56.0   NaN  58.0
2     2019-01-01 02:00  -6.5   NaN      60.0   NaN  44.0
3     2019-01-01 03:00  -6.9   NaN      62.0   NaN  48.0
4     2019-01-01 04:00  -7.2   NaN      57.0   NaN  51.0
...                ...   ...   ...       ...   ...   ...
8756  2019-12-31 20:00  -7.0   NaN      41.0   NaN  27.0
8757  2019-12-31 21:00  -7.4   NaN      43.0   NaN  27.0
8758  2019-12-31 22:00  -8.2   NaN      42.0   NaN  29.0
8759  2019-12-31 23:00  -7.4   NaN      45.0   NaN  24.0
8760  2020-01-01 00:00  -6.5   0.0      38.0   NaN  27.0

[8761 rows x 6 columns]

In [13]:
weather_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8761 entries, 0 to 8760
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      8761 non-null   object 
 1   temp      8761 non-null   float64
 2   rain      824 non-null    float64
 3   humidity  8761 non-null   float64
 4   snow      82 non-null     float64
 5   dust      8200 non-null   float64
dtypes: float64(5), object(1)
memory usage: 479.1+ KB


- date 변수의 변수 타입을 확인해보면, object라고 되어있음을 확인할 수 있다.
- 그런데 NS Shop+ 매출 실적 데이터 셋의 date 변수의 변수 타입은 datetime이다.
    - 따라서 **date 변수의 변수 타입을 datetime으로 변경**해주도록 하겠다.

In [14]:
weather_final['date'] = pd.to_datetime(weather_final['date'], format = '%Y-%m-%d %H:%M:%S')
weather_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8761 entries, 0 to 8760
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   date      8761 non-null   datetime64[ns]
 1   temp      8761 non-null   float64       
 2   rain      824 non-null    float64       
 3   humidity  8761 non-null   float64       
 4   snow      82 non-null     float64       
 5   dust      8200 non-null   float64       
dtypes: datetime64[ns](1), float64(5)
memory usage: 479.1 KB


In [15]:
# date 변수의 변수 타입이 datetime으로 잘 변경되었는지 확인
weather_final

date  temp  rain  humidity  snow  dust
0    2019-01-01 00:00:00  -5.5   NaN      54.0   NaN  59.0
1    2019-01-01 01:00:00  -5.9   NaN      56.0   NaN  58.0
2    2019-01-01 02:00:00  -6.5   NaN      60.0   NaN  44.0
3    2019-01-01 03:00:00  -6.9   NaN      62.0   NaN  48.0
4    2019-01-01 04:00:00  -7.2   NaN      57.0   NaN  51.0
...                  ...   ...   ...       ...   ...   ...
8756 2019-12-31 20:00:00  -7.0   NaN      41.0   NaN  27.0
8757 2019-12-31 21:00:00  -7.4   NaN      43.0   NaN  27.0
8758 2019-12-31 22:00:00  -8.2   NaN      42.0   NaN  29.0
8759 2019-12-31 23:00:00  -7.4   NaN      45.0   NaN  24.0
8760 2020-01-01 00:00:00  -6.5   0.0      38.0   NaN  27.0

[8761 rows x 6 columns]

**<이 부분은 일단 참고...>**
- 추가적으로 left join을 한 뒤, dust 변수가 NaN 값이었던 date가 언제인지 확인해보겠다.
    - 아래의 결과를 보면, **총 561개의 NaN 값들이 존재함**을 알 수 있다.
    - 즉, dust_var_select 데이터 셋에 아래의 date에 해당하는 dust 값들이 존재하지 않았었다는 것을 보여준다.
        - 8761개 - 561개 = 8200개

In [16]:
weather_final[weather_final['dust'].isnull()]

date  temp  rain  humidity  snow  dust
81   2019-01-04 09:00:00  -5.4   NaN      49.0   NaN   NaN
1106 2019-02-16 02:00:00  -2.9   NaN      69.0   0.5   NaN
1107 2019-02-16 03:00:00  -3.2   0.2      73.0   0.6   NaN
1108 2019-02-16 04:00:00  -3.5   NaN      72.0   0.7   NaN
1109 2019-02-16 05:00:00  -3.7   NaN      67.0   0.6   NaN
...                  ...   ...   ...       ...   ...   ...
8015 2019-11-30 23:00:00   5.4   NaN      50.0   NaN   NaN
8016 2019-12-01 00:00:00   5.3   NaN      51.0   NaN   NaN
8340 2019-12-14 12:00:00   3.3   NaN      49.0   NaN   NaN
8341 2019-12-14 13:00:00   4.4   NaN      44.0   NaN   NaN
8342 2019-12-14 14:00:00   5.9   NaN      39.0   NaN   NaN

[561 rows x 6 columns]

# 비 내림 여부 및 눈 내림 여부 변수 생성
- 현재 데이터 셋에 존재하는 rain, snow 변수에는 결측값들이 굉장히 많다.
- 따라서 **해당 일에 비 또는 눈이 한 시간대라도 내렸으면 "1", 그렇지 않으면 "0"**으로 처리해주는 **binary 변수**를 생성해서 사용하는 것이 추후에 모델의 예측 성능을 높이는 데 도움이 될 것으로 판단된다.
    - **비 내림 여부(is_rain)**
        - 0: 비 내리지 않음
        - 1: 비 내림
    - **눈 내림 여부(is_snow)**
        - 0: 눈 내리지 않음
        - 1: 눈 내림

- 앞서 말했듯이, rain 및 snow 변수에 굉장히 많은 결측값들이 존재한다.

In [17]:
weather_final.isnull().sum()

date           0
temp           0
rain        7937
humidity       0
snow        8679
dust         561
dtype: int64

- 먼저 date 변수를 활용하여 year, month, day 변수를 생성해주도록 하겠다.
    - 년/월/일 변수를 결합한 뒤(new_date), 새로 생성된 new_date 변수를 기준으로 비 내림 및 눈 내림 여부 변수를 만들어 주도록 하겠다.

In [18]:
weather_final['year'] = weather_final['date'].dt.year
weather_final['month'] = weather_final['date'].dt.month
weather_final['day'] = weather_final['date'].dt.day

weather_final

date  temp  rain  humidity  snow  dust  year  month  day
0    2019-01-01 00:00:00  -5.5   NaN      54.0   NaN  59.0  2019      1    1
1    2019-01-01 01:00:00  -5.9   NaN      56.0   NaN  58.0  2019      1    1
2    2019-01-01 02:00:00  -6.5   NaN      60.0   NaN  44.0  2019      1    1
3    2019-01-01 03:00:00  -6.9   NaN      62.0   NaN  48.0  2019      1    1
4    2019-01-01 04:00:00  -7.2   NaN      57.0   NaN  51.0  2019      1    1
...                  ...   ...   ...       ...   ...   ...   ...    ...  ...
8756 2019-12-31 20:00:00  -7.0   NaN      41.0   NaN  27.0  2019     12   31
8757 2019-12-31 21:00:00  -7.4   NaN      43.0   NaN  27.0  2019     12   31
8758 2019-12-31 22:00:00  -8.2   NaN      42.0   NaN  29.0  2019     12   31
8759 2019-12-31 23:00:00  -7.4   NaN      45.0   NaN  24.0  2019     12   31
8760 2020-01-01 00:00:00  -6.5   0.0      38.0   NaN  27.0  2020      1    1

[8761 rows x 9 columns]

In [19]:
# 아래의 for문을 돌리기 위해 임시적으로 "0" 값으로 채워진 new_date 변수를 미리 만들어 주었음
weather_final['new_date'] = 0

for i in tqdm(range(len(weather_final))):
    weather_final['new_date'][i] = str(weather_final['year'][i]) + '-' + str(weather_final['month'][i]) + '-' + str(weather_final['day'][i])
weather_final

C:\Users\Playdata\anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
C:\Users\Playdata\anaconda3\lib\site-packages\pandas\core\indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


date  temp  rain  humidity  snow  dust  year  month  day  \
0    2019-01-01 00:00:00  -5.5   NaN      54.0   NaN  59.0  2019      1    1   
1    2019-01-01 01:00:00  -5.9   NaN      56.0   NaN  58.0  2019      1    1   
2    2019-01-01 02:00:00  -6.5   NaN      60.0   NaN  44.0  2019      1    1   
3    2019-01-01 03:00:00  -6.9   NaN      62.0   NaN  48.0  2019      1    1   
4    2019-01-01 04:00:00  -7.2   NaN      57.0   NaN  51.0  2019      1    1   
...                  ...   ...   ...       ...   ...   ...   ...    ...  ...   
8756 2019-12-31 20:00:00  -7.0   NaN      41.0   NaN  27.0  2019     12   31   
8757 2019-12-31 21:00:00  -7.4   NaN      43.0   NaN  27.0  2019     12   31   
8758 2019-12-31 22:00:00  -8.2   NaN      42.0   NaN  29.0  2019     12   31   
8759 2019-12-31 23:00:00  -7.4   NaN      45.0   NaN  24.0  2019     12   31   
8760 2020-01-01 00:00:00  -6.5   0.0      38.0   NaN  27.0  2020      1    1   

        new_date  
0       2019-1-1  
1       2019-1-1  
2       2019-1-1  
3       2019-1-1  
4       2019-1-1  
...          ...  
8756  2019-12-31  
8757  2019-12-31  
8758  2019-12-31  
8759  2019-12-31  
8760    2020-1-1  

[8761 rows x 10 columns]

- 편리성을 위해 rain 및 snow 변수의 NaN 값들을 모두 "0"으로 바꿔준 뒤 작업을 시작하도록 하겠다.
    - 단, 기존의 rain 및 snow 변수는 건드리지 않고 rain_fillna, snow_fillna 변수를 새롭게 만들어서 사용하도록 하겠다.

In [20]:
weather_final['rain_fillna'] = weather_final['rain'].fillna(0)
weather_final['snow_fillna'] = weather_final['snow'].fillna(0)
weather_final.isnull().sum()

date              0
temp              0
rain           7937
humidity          0
snow           8679
dust            561
year              0
month             0
day               0
new_date          0
rain_fillna       0
snow_fillna       0
dtype: int64

In [21]:
# 시간대별(date)로 존재했던 강수량 및 적설량을 일자별(new_date)로 합쳐주기(sum) 
weather_pivot = pd.pivot_table(weather_final, index = 'new_date', aggfunc = np.sum)
# 비 내림 및 눈 내림 여부 생성을 위해 필요한 변수들만 선택 
weather_pivot = weather_pivot[['rain_fillna', 'snow_fillna']]
weather_pivot

rain_fillna  snow_fillna
new_date                           
2019-1-1           0.0          0.0
2019-1-10          0.0          0.0
2019-1-11          0.0          0.0
2019-1-12          0.0          0.0
2019-1-13          0.0          0.0
...                ...          ...
2019-9-6           2.4          0.0
2019-9-7           2.7          0.0
2019-9-8           0.1          0.0
2019-9-9           0.0          0.0
2020-1-1           0.0          0.0

[366 rows x 2 columns]

In [22]:
# 아래의 for문을 돌리기 위해 임시적으로 "0" 값으로 채워진 is_rain, is_snow 변수를 미리 만들어 주었음
weather_pivot['is_rain'] = 0
weather_pivot['is_snow'] = 0

for i in range(len(weather_pivot)):
    # 비가 내리지 않은 경우
    if weather_pivot['rain_fillna'][i] == 0:
        weather_pivot['is_rain'][i] = 0
    # 비가 내린 경우
    else:
        weather_pivot['is_rain'][i] = 1
        
    # 눈이 내리지 않은 경우
    if weather_pivot['snow_fillna'][i] == 0:
        weather_pivot['is_snow'][i] = 0
    # 눈이 내린 경우
    else:
        weather_pivot['is_snow'][i] = 1

C:\Users\Playdata\anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Playdata\anaconda3\lib\site-packages\ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  from ipykernel import kernelapp as app
C:\Users\Playdata\anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.in

In [23]:
weather_pivot['is_rain'].value_counts()

0    268
1     98
Name: is_rain, dtype: int64

In [24]:
weather_pivot['is_snow'].value_counts()

0    360
1      6
Name: is_snow, dtype: int64

In [25]:
df_is_rain_snow = weather_pivot.reset_index()
df_is_rain_snow = df_is_rain_snow[['new_date', 'is_rain', 'is_snow']]
df_is_rain_snow

new_date  is_rain  is_snow
0     2019-1-1        0        0
1    2019-1-10        0        0
2    2019-1-11        0        0
3    2019-1-12        0        0
4    2019-1-13        0        0
..         ...      ...      ...
361   2019-9-6        1        0
362   2019-9-7        1        0
363   2019-9-8        1        0
364   2019-9-9        0        0
365   2020-1-1        0        0

[366 rows x 3 columns]

- 이제 만들어진 df_is_rain_snow 데이터 셋과 기존의 weather_final 데이터 셋을 결합해주도록 하겠다.
    - new_date 변수를 조인키로 사용하여 **inner join**을 수행해주도록 하겠다.

In [26]:
weather_final2 = pd.merge(weather_final, df_is_rain_snow, how = 'inner', on = 'new_date')
weather_final2

date  temp  rain  humidity  snow  dust  year  month  day  \
0    2019-01-01 00:00:00  -5.5   NaN      54.0   NaN  59.0  2019      1    1   
1    2019-01-01 01:00:00  -5.9   NaN      56.0   NaN  58.0  2019      1    1   
2    2019-01-01 02:00:00  -6.5   NaN      60.0   NaN  44.0  2019      1    1   
3    2019-01-01 03:00:00  -6.9   NaN      62.0   NaN  48.0  2019      1    1   
4    2019-01-01 04:00:00  -7.2   NaN      57.0   NaN  51.0  2019      1    1   
...                  ...   ...   ...       ...   ...   ...   ...    ...  ...   
8756 2019-12-31 20:00:00  -7.0   NaN      41.0   NaN  27.0  2019     12   31   
8757 2019-12-31 21:00:00  -7.4   NaN      43.0   NaN  27.0  2019     12   31   
8758 2019-12-31 22:00:00  -8.2   NaN      42.0   NaN  29.0  2019     12   31   
8759 2019-12-31 23:00:00  -7.4   NaN      45.0   NaN  24.0  2019     12   31   
8760 2020-01-01 00:00:00  -6.5   0.0      38.0   NaN  27.0  2020      1    1   

        new_date  rain_fillna  snow_fillna  is_rain  is_snow  
0       2019-1-1          0.0          0.0        0        0  
1       2019-1-1          0.0          0.0        0        0  
2       2019-1-1          0.0          0.0        0        0  
3       2019-1-1          0.0          0.0        0        0  
4       2019-1-1          0.0          0.0        0        0  
...          ...          ...          ...      ...      ...  
8756  2019-12-31          0.0          0.0        0        0  
8757  2019-12-31          0.0          0.0        0        0  
8758  2019-12-31          0.0          0.0        0        0  
8759  2019-12-31          0.0          0.0        0        0  
8760    2020-1-1          0.0          0.0        0        0  

[8761 rows x 14 columns]

- 위 결과를 보면 알 수 있듯이, inner join이 잘 수행되었다.
- 이제 is_rain, is_snow 변수 생성을 위해 **임시로 만들어 놓은 "year, month, day, new_date, rain_fillna, snow_fillna" 변수들은 제거**해주도록 하겠다.

In [27]:
weather_final2.drop(['year', 'month', 'day', 'new_date', 'rain_fillna', 'snow_fillna'], axis = 1, inplace = True)
weather_final2

date  temp  rain  humidity  snow  dust  is_rain  is_snow
0    2019-01-01 00:00:00  -5.5   NaN      54.0   NaN  59.0        0        0
1    2019-01-01 01:00:00  -5.9   NaN      56.0   NaN  58.0        0        0
2    2019-01-01 02:00:00  -6.5   NaN      60.0   NaN  44.0        0        0
3    2019-01-01 03:00:00  -6.9   NaN      62.0   NaN  48.0        0        0
4    2019-01-01 04:00:00  -7.2   NaN      57.0   NaN  51.0        0        0
...                  ...   ...   ...       ...   ...   ...      ...      ...
8756 2019-12-31 20:00:00  -7.0   NaN      41.0   NaN  27.0        0        0
8757 2019-12-31 21:00:00  -7.4   NaN      43.0   NaN  27.0        0        0
8758 2019-12-31 22:00:00  -8.2   NaN      42.0   NaN  29.0        0        0
8759 2019-12-31 23:00:00  -7.4   NaN      45.0   NaN  24.0        0        0
8760 2020-01-01 00:00:00  -6.5   0.0      38.0   NaN  27.0        0        0

[8761 rows x 8 columns]

In [28]:
weather_final2['is_rain'].value_counts()

0    6409
1    2352
Name: is_rain, dtype: int64

In [29]:
weather_final2['is_snow'].value_counts()

0    8617
1     144
Name: is_snow, dtype: int64

# 최종 생성된 데이터 셋 csv 파일로 내보내기

In [30]:
weather_final2.to_csv('C:/Users/Playdata/2020_Bigcontest/dataset/weather/weather_final.csv', index = False)